File: app/src/main/java/mariavv/fitnesspal/data/repository/Repo.java
package mariavv.fitnesspal.data.repository;

import android.database.Cursor;

import java.util.Date;

import mariavv.fitnesspal.domain.Dish;
import mariavv.fitnesspal.domain.Food;

public class Repo {

    private static Repo instance;

    private DbRepository db;

    private Repo() {
        db = new DbRepository();
    }

    public static synchronized Repo getInstance() {
        if (instance == null) {
            instance = new Repo();
        }
        return instance;
    }

    public Cursor getFoodsFromHandbook() {
        return db.getFoodsFromHandbook();
    }

    public Cursor getJournal(long date) {
        return db.getJournal(date);
    }

    public void insertFoodInHandbook(Food food) {
        db.insertFoodInHandbook(food);
    }

    public int getJournalDaysCount() {
        final Cursor c = db.getJournalDaysCount();
        c.moveToFirst();
        final int count = c.getInt(0);
        c.close();
        return count;
    }

    public long getDateByIndex(int i) {
        final Cursor c = db.getJournalDates();
        if (i < c.getCount()) {
            c.moveToPosition(i);
            final long date = c.getLong(0);
            c.close();
            return date;
        } else {
            c.close();
            return new Date().getTime();
        }
    }

    public long insertDishInJournal(Dish dish) {
        return db.insertDishInJournal(dish);
    }

    public void clearJournal() {
        db.clearJournal();
    }

    public Cursor getFoodList() {
        return db.getFoodNamesFromHandbook();
    }

    public long addDish(Date date, String meal, String dish, int weight) {
        try {
            return db.insertDishInJournal(new Dish(date, meal, db.getFoodIdByName(dish), weight));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return -1;
    }

    public long addFood(Food food) {
        return db.insertFoodInHandbook(food);
    }

    public long[] getJournalDates() {
        final Cursor c = db.getJournalDates();
        final long[] dates = new long[c.getCount()];
        int i = 0;
        while (c.moveToNext()) {
            dates[i++] = c.getLong(0);
        }
        c.close();
        return dates;
    }
}

File: app/src/main/java/mariavv/fitnesspal/data/repository/Repo.kt
package mariavv.fitnesspal.data.repository

import android.database.Cursor
import mariavv.fitnesspal.domain.Dish
import mariavv.fitnesspal.domain.Food
import java.util.*

class Repo /*private constructor()*/ {

    private val db: DbRepository

    val foodsFromHandbook: Cursor
        get() = db.foodsFromHandbook

    val journalDaysCount: Int
        get() {
            val c = db.journalDaysCount
            c.moveToFirst()
            val count = c.getInt(0)
            c.close()
            return count
        }

    val foodList: Cursor
        get() = db.foodNamesFromHandbook

    val journalDates: LongArray
        get() {
            val c = db.journalDates
            val dates = LongArray(c.count)
            var i = 0
            while (c.moveToNext()) {
                dates[i++] = c.getLong(0)
            }
            c.close()
            return dates
        }

    init {
        db = DbRepository()
    }

    fun getJournal(date: Long): Cursor {
        return db.getJournal(date)
    }

    fun insertFoodInHandbook(food: Food) {
        db.insertFoodInHandbook(food)
    }

    fun getDateByIndex(i: Int): Long {
        val c = db.journalDates
        if (i > -1 && i < c.count) {
            c.moveToPosition(i)
            val date = c.getLong(0)
            c.close()
            return date
        } else {
            c.close()
            return Date().time
        }
    }

    fun insertDishInJournal(dish: Dish): Long {
        return db.insertDishInJournal(dish)
    }

    fun clearJournal() {
        db.clearJournal()
    }

    fun addDish(date: Date, meal: String, dish: String, weight: Int): Long {
        try {
            return db.insertDishInJournal(Dish(date, meal, db.getFoodIdByName(dish), weight))
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return -1
    }

    fun addFood(food: Food): Long {
        return db.insertFoodInHandbook(food)
    }

    companion object {

        val instance: Repo
            get() = Repo()


        /*@Synchronized
        fun getInstance(): Repo {
            if (instance == null) {
                instance = Repo()
            }
            return instance
        }*/
    }
}
